### DISCUSSION TITLE
LiDAR simulation parameters

### DISCUSSION THREAD

**Discussion by ashchennikov** [tud-cor/FS19_modROS#57] [Q&A]
I am trying to paramterize the laser scan in the mod_config.lua to reflect real LiDAR characteristics and seeing some strange things: I want to reduce the maximum and the minimum angles of the scan. If I set angle_min and angle_max to something other than -180° (-math.pi) and 180° (math.pi) respectively, the point clouds that I see in rviz get rotated or mirrored. Setting the rotation of the sensor does not help. I am not sure if I am doing something wrong. How do I parameterize this correctly to make the scan only "see" 90° in front of the machine?

I also want to increase the number of rays and layers to again reflect a real LiDAR. Are there any known relations between the amount of rays and the performance of a PC/graphics card? For example: The game is able to run smoothly with 128 * 5 rays on my setup (an office PC with a Nvidia Quadro P400). At 128 * 10 rays the framerate drops to about 15 FPS. Can it be said that the more rays I use the more powerful graphics card I need? Is it at all imaginable to use the matrix of let's say 300 * 100 rays?

**Comment by gavanderhoorn** [comment]
First: it's been a while since I've looked at the actual code. It's very much possible there are issues with it, especially some of the more 'math heavy' parts. We only tested an MVP as it turned out various aspects of FS (perhaps specific to FS19) made the use of it for our purposes less than ideal (causing us to abandon development).

> How do I parameterize this correctly to make the scan only "see" 90° in front of the machine?

the idea was indeed this should be possible by changing the lidar parameters in the sensor configuration.

If that doesn't immediately work, it's possible there is a bug in the implementation. Note there are two sides to this: the "FS19" side (as in: variables changing the way the Lua code generates the samples) and the parts of the code which configure the ROS `LaserScan` message. If those don't agree on the important aspects, ROS may interpret the data incorrectly, even though the FS19 side is working correctly.

I'd have to take a look to see what could be going wrong. If you happen to already have done that, please let me know. The implementation is in [utils/LaserScanner.lua](https://github.com/tud-cor/FS19_modROS/blob/869a170967091ee8a9ec7b833ad7aaa06d3f3828/src/utils/LaserScanner.lua).

> I also want to increase the number of rays and layers to again reflect a real LiDAR. Are there any known relations between the amount of rays and the performance of a PC/graphics card?

well .. nothing more specific than "increase nr of points -> decrease performance". We would not have anything more than that at the moment I'm afraid.

The 'laser scanner' is implemented using a rather naive approach using [raycastClosest(..)](https://github.com/tud-cor/FS19_modROS/blob/869a170967091ee8a9ec7b833ad7aaa06d3f3828/src/utils/LaserScanner.lua#L60). That's just a sw only (AFAIK) raycasting function provided by the FS19 game engine. It's mostly intended to be used for incidental calculation of distances between objects. Not for emulating sensors as we're using it in `modROS`.

> Is it at all imaginable to use the matrix of let's say 300 * 100 rays?

Without trying it out I can't really say, but my expectation would be: *no*. That's 30000 points, which even if sampling every point would only take 1 ms would already take 30 seconds. With a cost of 10 microseconds per point it would still be 0.3 seconds, which would still be too slow for 30 FPS (given there is more to do in a single frame than just running `modROS`).

We only went with `raycastClosest(..)` as we couldn't find any method of grabbing depth buffers (#22). It's far from an optimal way to do this.
